Feat: Reassign field ids for schema

c-thiel opened the pull request:
Required for `TableMetadataBuilder` (https://github.com/apache/iceberg-rust/pull/587).

When new TableMetadata is created, field ids should be reassign to ensure that field numbering is normalized and started from 0.

Xuanwo commented on crates/iceberg/src/spec/schema.rs:
```diff
@@ -86,6 +87,16 @@ impl SchemaBuilder {
+    /// If not provided, it will start from 0.
+    ///
+    /// All specified aliases and identifier fields will be updated to the new field-ids.
+    pub fn with_reassigned_field_ids(mut self, start_from: Option<i32>) -> Self {
```
Could you please elaborate on the circumstances under which the user wants to set this? I used to think users always need to start with `0`.

Xuanwo commented on crates/iceberg/src/spec/schema.rs:
```diff
@@ -86,6 +87,16 @@ impl SchemaBuilder {
+    ///
+    /// All specified aliases and identifier fields will be updated to the new field-ids.
+    pub fn with_reassigned_field_ids(mut self, start_from: Option<i32>) -> Self {
+        self.reassign_field_ids_from = Some(start_from.unwrap_or(0));
```
If we always set `Some(v)` here, how about just using `i32` for `reassign_field_ids_from` with default value `0`?

Fokko replied:
Typically you would pass in `last-column-id` from the table metadata. In the case of a `CREATE TABLE` this will be zero, in the case of a `REPLACE TABLE` it will be `last-column-id` where all columns get a new ID and you ensure they are not used before.

c-thiel replied:
@Xuanwo only doubt with the i32 is that people might input 1 as something is required - and it's shorter than Default::default().
The rust tests are quite a good example that this can happen: We have many tests where field numbering starts from 1, so I had to re-write them to be compatible with the new builder.

I am OK with the i32 as well, just wanted to bring up that a None is probably more verbose.

c-thiel replied:
Both good points, fixed in the latest commit.
Its `dead_code` for now

Xuanwo commented:
cc @liurenjie1024 would you like to take a look too?

c-thiel commented on crates/iceberg/src/spec/schema.rs:
```diff
@@ -943,6 +965,122 @@ impl SchemaVisitor for PruneColumn {
+        let outer_fields = fields
+            .into_iter()
+            .map(|field| {
+                self.old_to_new_id.insert(field.id, self.next_field_id);
```
I added the check. It should never trigger currently, because I switched the build method order as you suggested.
Nevertheless the check is good if Reassignment is ever used outside of the builder in the future.

c-thiel commented on crates/iceberg/src/spec/schema.rs:
```diff
@@ -105,7 +116,16 @@ impl SchemaBuilder {
     /// Builds the schema.
-    pub fn build(self) -> Result<Schema> {
+    pub fn build(mut self) -> Result<Schema> {
+        if let Some(start_from) = self.reassign_field_ids_from {
```
I agree - I didn't think about the error messages.
It's unfortunately not super straightforward to switch the order, as for example index_by_id needs a full rebuild as both keys and values changes. I think we have two options to achieve the swap:

1. Use `id_reassigner.apply_*` for all other fields that have been build (i.e. `id_to_field`) or
2. Re-build those fields using the same code - i.e. the build option.

I opted for 2.
While it is slightly less performant because we have to visit the structure again, it is much less error prone as we simply re-use existing and well-tested code.

Let me know what you think!

c-thiel replied:
@liurenjie1024 I refactored it in my latest commit, however we cannot get rid of the two functions:
`apply_to_aliases` and `apply_to_identifier_fields`. The builder itself is standard now.

c-thiel commented:
@liurenjie1024 I addressed all of your comments.
I also noticed that I didn't handle a potential field_id overflow error which I also fixed.

Regarding your uniqueness of id comment: I noticed that this wasn't enforced previously, so I fixed it and added a test for it as well. This is currently making a few tests crash, where key-ids & value-ids of maps are re-used elsewhere in the schema.

I am checking how java handles this and will fix the tests accordingly.

c-thiel commented:
Ok, java does not allow this. It's easy to test with the `testNestedStructSchemaToMapping` java test.
I fixed the rusts test that failed due to duplicate nested IDs.

@liurenjie1024  ready for another round :)

c-thiel commented on crates/iceberg/src/spec/schema.rs:
```diff
@@ -86,6 +87,16 @@ impl SchemaBuilder {
         self
     }
 
+    /// Reassign all field-ids (nested) on build.
```
Fixed in latest commit

c-thiel commented:
@liurenjie1024 last two minor changes added - I think we are good to merge @liurenjie1024  @Xuanwo
